providers/supermicro/firmware.go

```diff
@@ -3,104 +3,119 @@ package supermicro
-	default:
-		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstall, "component unsupported: "+component)
+func (c *Client) FirmwareUpload(ctx context.Context, component string, reader io.Reader) (taskID string, err error) {
+	if err := c.firmwareInstallSupported(ctx); err != nil {
+		return "", err
 	}
 
-	if err != nil {
-		err = errors.Wrap(bmclibErrs.ErrFirmwareInstall, err.Error())
+	//	// expect atleast 5 minutes left in the deadline to proceed with the upload
+	d, _ := ctx.Deadline()
+	if time.Until(d) < 5*time.Minute {
+		return "", errors.New("remaining context deadline insufficient to perform update: " + time.Until(d).String())
 	}
 
-	return jobID, err
-}
+	switch {
+	case strings.HasPrefix(strings.ToLower(c.model), "x12"):
+		return c.x12().firmwareUpload(ctx, component, reader)
+	case strings.HasPrefix(strings.ToLower(c.model), "x11"):
+		return c.x11().firmwareUpload(ctx, component, reader)
+	}
 
-// FirmwareInstallStatus returns the status of the firmware install process
-func (c *Client) FirmwareInstallStatus(ctx context.Context, installVersion, component, taskID string) (string, error) {
-	component = strings.ToUpper(component)
+	return "", errors.Wrap(errUnexpectedModel, c.model)
+}
 
-	switch component {
-	case common.SlugBMC:
-		return c.statusBMCFirmwareInstall(ctx)
-	case common.SlugBIOS:
-		return c.statusBIOSFirmwareInstall(ctx)
-	default:
-		return "", errors.Wrap(bmclibErrs.ErrFirmwareInstallStatus, "component unsupported: "+component)
+func (c *Client) FirmwareInstallUploaded(ctx context.Context, component, uploadTaskID string) (installTaskID string, err error) {
+	if err := c.firmwareInstallSupported(ctx); err != nil {
+		return "", err
 	}
-}
 
-func (c *Client) deviceSupported(ctx context.Context) error {
-	errBoardPartNumUnknown := errors.New("baseboard part number unknown")
-	errBoardUnsupported := errors.New("feature not supported/implemented for device")
+	if uploadTaskID == "" {
+		return "", errUploadTaskIDExpected
+	}
 
-	// Its likely this works on all X11's
-	// for now, we list only the ones its been tested on.
-	//
-	// board part numbers
-	//
-	supported := []string{
-		"X11SCM-F",
-		"X11DPH-T",
-		"X11SCH-F",
-		"X11DGQ",
-		"X11DPG-SN",
-		"X11DPT-B",
-		"X11SSE-F",
+	switch {
+	case strings.HasPrefix(strings.ToLower(c.model), "x12"):
+		return c.x12().firmwareInstallUploaded(ctx, component, uploadTaskID)
+	case strings.HasPrefix(strings.ToLower(c.model), "x11"):
+		return "", c.x11().firmwareInstallUploaded(ctx, component)
 	}
 
-	data, err := c.fruInfo(ctx)
-	if err != nil {
-		return err
+	return "", errors.Wrap(errUnexpectedModel, c.model)
+
+}
+
+// FirmwareTaskStatus returns the status of a firmware related task queued on the BMC.
+func (c *Client) FirmwareTaskStatus(ctx context.Context, kind constants.FirmwareInstallStep, component, taskID, installVersion string) (state, status string, err error) {
+	if err := c.firmwareInstallSupported(ctx); err != nil {
```

Comment:
This is because of the way these bmc client interface methods are called - `FirmwareUpload`, `FirmwareTaskStatus` - each method is attempted on every provider (dell, redfish, smc...) and so each provider implementation has to make sure its running on supported hardware before proceeding (atleast I'd like to be twice as careful for firmware related actions).